services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/update/AddUpdateAction.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb.update;
+
+import static java.util.Collections.unmodifiableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A representation of a single {@link UpdateExpression} ADD action.
+ * <p>
+ * At a minimum, this action must contain a path string referencing the attribute that should be acted upon and a value string
+ * referencing the value to be added to the attribute. The value should be substituted with tokens using the
+ * ':value_token' syntax and values associated with the token must be explicitly added to the expressionValues map.
+ * Consult the DynamoDB UpdateExpression documentation for details on this action.
+ * <p>
+ * Optionally, attribute names can be substituted with tokens using the '#name_token' syntax. If tokens are used in the
+ * expression then the names associated with those tokens must be explicitly added to the expressionNames map
+ * that is also stored on this object.
+ * <p>
+ * Example:-
+ * <pre>
+ * {@code
+ * AddUpdateAction addAction = AddUpdateAction.builder()
+ *                                            .path("#a")
+ *                                            .value(":b")
+ *                                            .putExpressionName("#a", "attributeA")
+ *                                            .putExpressionValue(":b", myAttributeValue)
+ *                                            .build();
+ * }
+ * </pre>
+ */
+@SdkPublicApi
+public final class AddUpdateAction implements UpdateAction {
+
+    private final String path;
+    private final String value;
+    private final Map<String, String> expressionNames;
+    private final Map<String, AttributeValue> expressionValues;
+
+    private AddUpdateAction(Builder builder) {
+        this.path = Validate.paramNotNull(builder.path, "path");
+        this.value = Validate.paramNotNull(builder.value, "value");
+        this.expressionValues = unmodifiableMap(Validate.paramNotNull(builder.expressionValues, "expressionValues"));
```

Comment:
Is it true that this must always be non-null?

Reply:
AFAIK; not sure if you're just referring to expressionValues or all of them. For expressionValues I believe so, but we should test this premise before release. I figured it's better to be restrictive at first.